### DISCUSSION TITLE
Failing to run on MacOS

### DISCUSSION THREAD

**Discussion by til-git** [Genesis-Embodied-AI/genesis-world#175] [Q&A]
I'm trying to run the examples on a M1 Studio, but it fails in various forms.  

Using the render to file example for headless rendering [here](https://genesis-world.readthedocs.io/en/latest/user_guide/getting_started/visualization.html) it hangs after the first frame with this 

`  File "/Volumes/Home/tilman/Dev/genesis/render_to_file.py", line 55, in <module>
    cam.render()
  File "/opt/homebrew/lib/python3.11/site-packages/genesis/utils/misc.py", line 48, in wrapper
    return method(self, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/lib/python3.11/site-packages/genesis/vis/camera.py", line 165, in render
    rgb_arr, depth_arr, seg_idxc_arr, normal_arr = self._rasterizer.render_camera(
                                                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/lib/python3.11/site-packages/genesis/vis/rasterizer.py", line 57, in render_camera
    rgb_arr, depth_arr = self._viewer._pyrender_viewer.render_offscreen(
                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/lib/python3.11/site-packages/genesis/ext/pyrender/viewer.py", line 588, in render_offscreen
    self._offscreen_result_semaphore.acquire()
  File "/opt/homebrew/Cellar/python@3.11/3.11.11/Frameworks/Python.framework/Versions/3.11/lib/python3.11/threading.py", line 479, in acquire
    self._cond.wait(timeout)
  File "/opt/homebrew/Cellar/python@3.11/3.11.11/Frameworks/Python.framework/Versions/3.11/lib/python3.11/threading.py", line 327, in wait
    waiter.acquire()
`

When trying to run with the macos specific example [render_on_macos.py](https://github.com/Genesis-Embodied-AI/Genesis/blob/main/examples/render_on_macos.py) it runs through the frames, but no window pops up.  

This with no reported package incompatibilities, python3.11, MacOS 15.1.1.  No errors during startup or rendering are reported. 

Any help in getting off the ground on MacOS highly appreciated.

**Comment by analogjedi** [comment]
Hi @til-git . Many users are experiencing the same or similar issues with MacOS and Apple Silicon. There are several issues logged. Nothing has worked for me yet. I gave up after battling with it for 1/2 a day and will circle back soon.

**Reply by til-git** [reply]
Thanks @analogjedi, found the bugs and tracking them - some claim success, but most seem to fail in GL stuff.

**Comment by TwinSnakes007** [comment]
Genesis is up and running on M# Apple Silicon machines - after some headaches.

I can post what worked for me if needed.

**Reply by til-git** [reply]
Hi @TwinSnakes007 - yes, would love to know what worked for you.

**Reply by TwinSnakes007** [reply]
Assumes Homebrew is already installed

brew install python@3.12
brew install ffmpeg
brew install pixi
pixi init
pixi add python=3.12.8
pixi add pip
pixi shell
pip install genesis-world
pip install torch torchvision torchaudio

**Reply by chriswaligorski** [reply]
Did this also let the viewer pop up?

**Reply by andrewlyu** [reply]
viewer does not pop up on m1 macbook

**Comment by til-git** [comment]
Yeah, same here - no viewer and still getting the offscreen semaphore locking issue, even installing exactly as described.

**Reply by TwinSnakes007** [reply]
Since the docker has been published.  I'm going to try that on MacOS and see how it fares.

Dunno if anyone has already tried that pathway?

**Reply by andrewlyu** [reply]
The following configuration works for me finally:
1. anaconda
2. python3.11
3. [Very Important] hardcode args.vis = True and args.mac = True
It looks the argument parser needs a fix on macos.

BTW, docker was not successfully built on my m1 macbook.

**Reply by TwinSnakes007** [reply]
Docker didn't compile for me either on M2 Sequoia...it failed compiling the ray tracer.

**Comment by nickarino** [comment]
Is there a working Docker image on dockerhub for the macos ARM64?

**Comment by nickarino** [comment]
@andrewlyu, I am on a M3, what would you change about this script? 
`````
# Pull the latest changes from the remote repository
git pull origin main

# Ensure the correct Python version is used and create a virtual environment
~/.pyenv/versions/3.10.12/bin/python -m venv venv

# Activate the virtual environment
source venv/bin/activate

# Verify the Python version
python --version

# Install PyTorch and related packages using CPU-compatible wheels
pip install torch torchvision torchaudio --index-url https://download.pytorch.org/whl/cpu                

# Install the project as an editable package
pip install -e .

pip install genesis-world

#install missing parts

brew install ffmpeg python-tk

source venv/bin/activate

pip install --upgrade pip

pip install --upgrade taichi pygel3d trimesh pyrender
export IMAGEIO_FFMPEG_EXE=$(which ffmpeg)
export TK_SILENCE_DEPRECATION=1

git submodule update --init --recursive

cd examples

python render_on_macos.py

```

**Comment by kevenzhou** [comment]
my mac is m2, run with --vis can fix
